File: java/01653_minimum_deletions_to_make_string_balanced.java
//        Java
// Time complexity - O (N)
// Space complexity - O (1)
class Solution {
    public int minimumDeletions(String s) {
        int bCount = 0;
        int deletions = 0;

        for (char c : s.toCharArray()) {
            if (c == 'b') {
                bCount++;
            } else {
                deletions = Math.min(deletions + 1, bCount);
            }
        }

        return deletions;
    }
}

File: python/01653_minimum_deletions_to_make_string_balanced.py
#        Python
# Time complexity - O (n)
# Space complexity - O (1)
class Solution:

    # "aababbab" => "aaabbb" => 2

    # "aaaabbbbaaaaaaa"
    def minimumDeletions(self, s: str) -> int:
        b = 0
        delete = 0

        for char in s:
            if char == 'b':
                b += 1
            else:
                delete += 1
                delete = min(delete, b) # aaaabbbb,  aaaabbbbaaa, aaaabbbbaaaaaaa

        return delete
